File: leetcode/46.permutations.py
# @leet start
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        res: list[list[int]] = []

        def backtrack(i: int) -> None:
            if i == len(nums):
                res.append(nums[:])

            for j in range(i, len(nums)):
                nums[i], nums[j] = nums[j], nums[i]
                backtrack(i + 1)
                nums[i], nums[j] = nums[j], nums[i]

        backtrack(0)

        return res


# @leet end

File: leetcode/46.permutations.ts
// @leet start
function permute(nums: number[]): number[][] {
  const result: number[][] = [];

  const backtrack = (start: number) => {
    if (start === nums.length - 1) {
      result.push([...nums]);
    }

    for (let i = start; i < nums.length; i++) {
      [nums[start], nums[i]] = [nums[i], nums[start]];
      backtrack(start + 1);
      [nums[start], nums[i]] = [nums[i], nums[start]];
    }
  };

  backtrack(0);

  return result;
}
// @leet end
